x-pack/plugins/security/public/management/api_keys/api_keys_grid/api_keys_grid_page.tsx

```diff
@@ -33,35 +22,108 @@ import { FormattedMessage } from '@kbn/i18n-react';
 import { KibanaPageTemplate } from '@kbn/shared-ux-page-kibana-template';
 import { Route } from '@kbn/shared-ux-router';
-import { UserAvatar, UserProfilesPopover } from '@kbn/user-profile-components';
 
 import { ApiKeyFlyout } from './api_key_flyout';
 import { ApiKeysEmptyPrompt } from './api_keys_empty_prompt';
+import { ApiKeysTable, MAX_PAGINATED_ITEMS } from './api_keys_table';
+import type { CategorizedApiKey } from './api_keys_table';
 import { InvalidateProvider } from './invalidate_provider';
-import type { ApiKey, AuthenticatedUser, RestApiKey } from '../../../../common';
+import type { ApiKey } from '../../../../common/model';
 import { Breadcrumb } from '../../../components/breadcrumb';
 import { SelectableTokenField } from '../../../components/token_field';
 import { useCapabilities } from '../../../components/use_capabilities';
 import { useAuthentication } from '../../../components/use_current_user';
-import type { CreateAPIKeyResult } from '../api_keys_api_client';
+import type { CreateAPIKeyResult, QueryApiKeySortOptions } from '../api_keys_api_client';
 import { APIKeysAPIClient } from '../api_keys_api_client';
 
+const parseSearchBarQuery = (query: Query): QueryContainer => {
+  let parsedQuery = query;
+
+  if (parsedQuery.hasSimpleFieldClause('owner')) {
+    const ownerQueryValue = parsedQuery.getSimpleFieldClause('owner')!.value;
+    parsedQuery = parsedQuery
+      .removeSimpleFieldClauses('owner')
+      .addSimpleFieldValue('username', `${ownerQueryValue}`);
+  }
+
+  if (query.text.includes('type:managed')) {
+    const subQuery = query.text.replace('type:managed', '');
+    parsedQuery = EuiSearchBar.Query.parse(`${subQuery} (metadata.managed:true OR Alerting*)`);
+  } else if (query.text.includes('type:rest') || query.text.includes('type:cross_cluster')) {
+    parsedQuery = EuiSearchBar.Query.parse(`${query.text} -metadata.managed:true -Alerting*`);
+  }
+
+  if (query.text.includes('expired:true')) {
+    const subQuery = query.text.replace('expired:true', '');
+    parsedQuery = EuiSearchBar.Query.parse(`${subQuery} expiration<=${moment.now()}`);
+  } else if (query.text.includes('expired:false')) {
+    const subQuery = query.text.replace('expired:false', '');
+    parsedQuery = EuiSearchBar.Query.parse(`${subQuery} expiration>${moment.now()}`);
+  }
+
+  parsedQuery = parsedQuery.addSimpleFieldValue('invalidated', false);
+
+  return EuiSearchBar.Query.toESQuery(parsedQuery);
+};
+
 export const APIKeysGridPage: FunctionComponent = () => {
   const { services } = useKibana<CoreStart>();
   const history = useHistory();
   const authc = useAuthentication();
-  const [state, getApiKeys] = useAsyncFn(
-    () => Promise.all([new APIKeysAPIClient(services.http).getApiKeys(), authc.getCurrentUser()]),
-    [services.http]
-  );
 
+  const [query, setQuery] = useState<Query>(EuiSearchBar.Query.parse(''));
+  const [from, setFrom] = useState<number>(0);
+  const [pageSize, setPageSize] = useState(25);
```

Comment:
__issue:__ something strange is going on with paging

[Screencast from 2024-04-16 18-09-48.webm](https://github.com/elastic/kibana/assets/1713708/e3c168a3-dd9c-4374-9cc6-d4e3bbbdb3b9)

Reply:
So the issue here was that I was using the "total keys" present to show pagination even if the result of the query was different. Fixed in https://github.com/elastic/kibana/pull/168970/commits/b8aa5bf63277712f6c2ecb1ee450881b36f1281c